README.md
[WIP] Flask-sqla2api
==============

Flask middleware that creates a simple Flask API CRUD REST endpoint
based on a SQLAlchemy model definition.

#### Basic usage

```python
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_sqla2api import SQLA2api

# Init app and DB
app = Flask(__name__)
db = SQLAlchemy(app)

# Setup a simple SQLAlchemy model
class Entry(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(80))

# Init Flask-sqla2api and register its blueprint
api = SQLA2api(app, Entry, db)
app.register_blueprint(api.make_blueprint(), url_prefix='/')
```
The previous tiny app will create the following endpoints

| URL           | HTTP Method | Action                |
|---------------|-------------|-----------------------|
| `/entry`      | GET         | Get all entries       |
| `/entry`      | POST        | Create new entry      |
| `/entry/<id>` | GET         | Get a single entry    |
| `/entry/<id>` | PUT         | Edit existing entry   |
| `/entry/<id>` | DELETE      | Delete existing entry |

### TODO

* [x] Add PUT/GET/DELETE for single entry
* [ ] Add documentation
* [ ] Add various CI files (`.travis.yml`, `tox.ini`)
* [ ] Add `setup.py`
* [ ] Separate requirements for dev environments


CHANGES.rst
Changes
=======

All notable changes to this project will be documented in this file.

0.2.1 - 2018-02-20
------------------
- Better support for JSON input and output

0.2.0 - 2018-01-27
------------------
- Add multiple model support

0.1.0 - 2017-09-29
------------------
- First release


tests/test_api.py
import unittest

from flask import Flask, json
from flask_sqlalchemy import SQLAlchemy

from flask_sqla2api import SQLA2api

GOOD_ENTRY = {'id': 1, 'name': 'entry mcentryface'}
HEADERS = {'Content-Type': 'multipart/form-data'}


class ApiTests(unittest.TestCase):
    def setUp(self):
        app = Flask(__name__)
        app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
        app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///:memory:'
        self.db = SQLAlchemy(app)

        class Entry(self.db.Model):
            id = self.db.Column(self.db.Integer, primary_key=True)
            name = self.db.Column(self.db.String(80))

        class Person(self.db.Model):
            id = self.db.Column(self.db.Integer, primary_key=True)
            name = self.db.Column(self.db.String(80))

        api = SQLA2api([Entry, Person], self.db)
        api.append_blueprints(app)

        self.db.create_all()
        self.client = app.test_client()

    def tearDown(self):
        self.db.session.remove()
        self.db.drop_all()

    def test__get_all(self):
        self.client.post('/entry', data=GOOD_ENTRY, headers=HEADERS)
        res = self.client.get('/entry')
        json_res = json.loads(res.data)

        self.assertEqual(200, res.status_code)
        self.assertEqual(1, len(json_res))
        self.assertEqual(GOOD_ENTRY, json_res[0])

    def test__get_one(self):
        self.client.post('/entry', data=GOOD_ENTRY, headers=HEADERS)
        res = self.client.get('/entry/1')
        json_res = json.loads(res.data)

        self.assertEqual(200, res.status_code)
        self.assertEqual(GOOD_ENTRY, json_res)

    def test__get_one__not_found(self):
        res = self.client.get('/entry/2')
        self.assertEqual(404, res.status_code)

    def test__post(self):
        res = self.client.post('/entry', data=GOOD_ENTRY, headers=HEADERS)
        json_res = json.loads(res.data)

        self.assertEqual(201, res.status_code)
        self.assertEqual(GOOD_ENTRY, json_res)

    def test__post__no_content_type__throws(self):
        res = self.client.post('/entry', data=GOOD_ENTRY)
        self.assertEqual(415, res.status_code)

    def test__put(self):
        self.client.post('/entry', data=GOOD_ENTRY, headers=HEADERS)
        modified_entry = GOOD_ENTRY
        modified_entry['name'] = 'modified'
        res = self.client.put('/entry/1', data=modified_entry, headers=HEADERS)
        json_res = json.loads(res.data)

        self.assertEqual(201, res.status_code)
        self.assertEqual(modified_entry, json_res)

    def test__put__not_found(self):
        res = self.client.put('/entry/2', data=GOOD_ENTRY, headers=HEADERS)
        self.assertEqual(404, res.status_code)

    def test__delete(self):
        self.client.post('/entry', data=GOOD_ENTRY, headers=HEADERS)
        res = self.client.delete('/entry/1')

        self.assertEqual(204, res.status_code)
        self.assertEqual(b'', res.data)

    def test__delete__not_found(self):
        res = self.client.delete('/entry/2')
        self.assertEqual(404, res.status_code)


tests/test_raises.py
import unittest

from flask import Flask
from flask_sqlalchemy import SQLAlchemy

from flask_sqla2api import SQLA2api, generate_blueprint


class RaisesTest(unittest.TestCase):
    def setUp(self):
        self.api = SQLA2api()
        self.app = Flask(__name__)

    def test__append_blueprints__no_app__throws(self):
        with self.assertRaises(ValueError):
            self.api.append_blueprints(None)

    def test__generate_blueprint__no_model__throws(self):
        db = SQLAlchemy(self.app)
        with self.assertRaises(ValueError):
            generate_blueprint(None, db)

    def test__generate_blueprint__no_db__throws(self):
        db = SQLAlchemy(self.app)

        class TestModel(db.Model):
            id = db.Column(db.Integer, primary_key=True)

        with self.assertRaises(ValueError):
            generate_blueprint(TestModel, None)


flask_sqla2api/sqla2api.py
from flask import request
from .models import Model


class SQLA2api(object):
    def __init__(self, models=None, db=None):
        self.db = None
        self.models = []
        if db is not None:
            self.init_db(db)
        if models is not None:
            self.init_models(models)

    def init_models(self, models):
        self.models = []
        for model in models:
            self.models.append(Model(model, self.db))

    def init_db(self, db):
        self.db = db

    def append_blueprints(self, app):
        if app is None:
            raise ValueError("Cannot append to null app.")
        for model in self.models:
            app.register_blueprint(model.make_blueprint(), url_prefix='/')


def generate_blueprint(model, db):
    model = Model(model, db)
    return model.make_blueprint()


flask_sqla2api/models.py
from flask import Blueprint, jsonify, request


class Model(object):
    def __init__(self, model=None, db=None):
        self.model = None
        self.db = None
        if model is not None:
            self.init_model(model)
        if db is not None:
            self.init_db(db)

    def init_db(self, db):
        self.db = db

    def init_model(self, model):
        if model is None:
            raise ValueError("Invalid value None for model init")
        self.model = model
        self.alias = str(model.__table__)
        self.fields = model.__table__.columns.keys()
        self.fields_type = {}
        for column in model.__table__.columns:
            self.fields_type[column.name] = column.type.python_type

    def make_blueprint(self):
        if self.model is None or self.db is None:
            raise ValueError("Model or db not initiated.")
        blueprint = Blueprint(self.alias, __name__)

        @blueprint.route(self.alias, methods=['GET', 'POST'])
        def collection_methods():
            if request.method == 'POST':
                data = self.get_data(request)
                if data is None:
                    return "Unsupported media type", 415
                return self.post(data)
            else:
                return self.get_all()

        @blueprint.route(self.alias + '/<item_id>', methods=['GET', 'PUT', 'DELETE'])
        def resource_methods(item_id):
            entry = self.model.query.get_or_404(item_id)
            if request.method == 'GET':
                return self.get(entry)
            elif request.method == 'PUT':
                data = self.get_data(request)
                if data is None:
                    return "Unsupported media type", 415
                return self.put(data, entry)
            else:
                return self.delete(entry)

        return blueprint

    def post(self, data):
        params = {}
        for field in self.fields:
            params[field] = self.cast(field, data[field])
        new_item = self.model(**params)
        # TODO: add try/except
        self.db.session.add(new_item)
        self.db.session.commit()

        return jsonify(params), 201

    def get_all(self):
        all_items = self.model.query.all()
        results = []
        for item in all_items:
            entry = {}
            for field in self.fields:
                entry[field] = getattr(item, field)
            results.append(entry)

        return jsonify(results), 200

    def get(self, entry):
        result = {}
        for field in self.fields:
            result[field] = getattr(entry, field)

        return jsonify(result), 200

    def put(self, data, entry):
        result = {}
        for field in self.fields:
            value = self.cast(field, data[field])
            setattr(entry, field, value)
            result[field] = value
        self.db.session.commit()

        return jsonify(result), 201

    def delete(self, entry):
        self.db.session.delete(entry)
        self.db.session.commit()
        return '', 204

    def cast(self, field, value):
        if not self.fields_type:
            raise ValueError("Model not initialized")
        if value is None:
            raise ValueError(
                "Expected value but found None for field: %s" % field)
        return (self.fields_type[field])(value)

    def get_data(self, request):
        if "multipart/form-data" in request.content_type:
            return request.form.to_dict()
        if request.content_type == "application/json":
            return request.get_json()
        return
